things/service.go

```diff
@@ -140,50 +141,68 @@ func (svc service) ViewClientPerms(ctx context.Context, token, id string) ([]str
 	return permissions, nil
 }
 
-func (svc service) ListClients(ctx context.Context, token, reqUserID string, pm mgclients.Page) (mgclients.ClientsPage, error) {
-	var ids []string
-
+func (svc service) ListClients(ctx context.Context, token string, pm mgclients.Page) (mgclients.ClientsPage, error) {
 	res, err := svc.identify(ctx, token)
 	if err != nil {
 		return mgclients.ClientsPage{}, err
 	}
 
-	switch {
-	case (reqUserID != "" && reqUserID != res.GetUserId()):
-		// Check user is admin of domain, if yes then show listing on domain context
-		if _, err := svc.authorize(ctx, "", auth.UserType, auth.UsersKind, res.GetId(), auth.AdminPermission, auth.DomainType, res.GetDomainId()); err != nil {
-			return mgclients.ClientsPage{}, err
-		}
-		rtids, err := svc.listClientIDs(ctx, auth.EncodeDomainUserID(res.GetDomainId(), reqUserID), pm.Permission)
-		if err != nil {
-			return mgclients.ClientsPage{}, errors.Wrap(svcerr.ErrNotFound, err)
-		}
-		ids, err = svc.filterAllowedThingIDs(ctx, res.GetId(), pm.Permission, rtids)
-		if err != nil {
-			return mgclients.ClientsPage{}, errors.Wrap(svcerr.ErrNotFound, err)
-		}
-	default:
-		err := svc.checkSuperAdmin(ctx, res.GetUserId())
+	switch pm.EntityType {
+	case auth.UserType:
+		pm.EntityID = auth.EncodeDomainUserID(res.GetDomainId(), pm.EntityID)
+		return svc.listEntityThings(ctx, res.GetId(), res.GetDomainId(), auth.ViewPermission, auth.DomainType, res.GetDomainId(), pm)
+	case auth.GroupType:
+		pm.Permission = auth.GroupRelation
+		return svc.listEntityThings(ctx, res.GetId(), res.GetDomainId(), auth.ViewPermission, auth.GroupType, pm.EntityID, pm)
+	case "":
+		_, err := svc.authorize(ctx, "", auth.UserType, auth.UsersKind, res.GetId(), auth.ViewPermission, auth.DomainType, res.GetDomainId())
 		switch {
 		case err == nil:
 			pm.Domain = res.GetDomainId()
 		default:
-			// If domain is disabled , then this authorization will fail for all non-admin domain users
-			if _, err := svc.authorize(ctx, "", auth.UserType, auth.UsersKind, res.GetId(), auth.MembershipPermission, auth.DomainType, res.GetDomainId()); err != nil {
-				return mgclients.ClientsPage{}, err
-			}
-			ids, err = svc.listClientIDs(ctx, res.GetId(), pm.Permission)
-			if err != nil {
-				return mgclients.ClientsPage{}, errors.Wrap(svcerr.ErrNotFound, err)
-			}
+			pm.EntityType = auth.UserType
+			pm.EntityID = res.GetId()
+			return svc.listEntityThings(ctx, res.GetId(), res.GetDomainId(), auth.MembershipPermission, auth.DomainType, res.GetDomainId(), pm)
```

Comment:
If domina is disabled , then this should not work for domain non adminsitrator users. Please check this case

Reply:
I have checked this condition. When the domain is disabled only admins are able to list things